File: spring_lottoresult/src/main/java/com/eastgrams/lottoresult/service/LottoDrawResultService.java
package com.eastgrams.lottoresult.service;

import com.eastgrams.lottoresult.data.model.LottoDrawResult;
import com.eastgrams.lottoresult.data.repository.LottoDrawResultRepository;
import com.eastgrams.lottoresult.service.external.lottoresult.LottoDrawResultExternalService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
@RequiredArgsConstructor
public class LottoDrawResultService {
    private final LottoDrawResultRepository lottoDrawResultRepository;
    private final LottoDrawResultExternalService lottoDrawResultExternalService;

    public Optional<LottoDrawResult> getByDrwNo(int drwNo){

        Optional<LottoDrawResult> result = lottoDrawResultRepository.findByDrwNo(drwNo);
        if (result.isPresent()) {
            return result;
        }

        result = lottoDrawResultExternalService.getLottoDrawResult(drwNo);
        if(result.isPresent()){
            lottoDrawResultRepository.save(result.get());
            return result;
        }
        return Optional.empty();
    }
}

File: streamlit_lottoResult/app/service/lotto_draw_result_service.py
from typing import Optional
import threading

from app.data.model import LottoDrawResult
from app.system.utils import lotto_utils
from app.data.repository import LottoDrawResultRepository
import app.service.external as external

class LottoDrawResultService:

    def __init__(self, repo: LottoDrawResultRepository):
        self.repo = repo

    def _add_lotto_draw_result(self, result: LottoDrawResult) -> None:
        self.repo.save(result)

    async def get_lotto_draw_result(self, drw_no: int) -> Optional[LottoDrawResult]:
        result = self.repo.get_by_drw_no(drw_no)
        if result == None:
            result = await external.get_lotto_draw_result(drw_no)
            if result != None:
                add_thread = threading.Thread(target= self._add_lotto_draw_result, args=(result,))
                add_thread.start()
                return result
        return result